plugins/net_plugin/net_plugin.cpp

```diff
@@ -1256,58 +1254,66 @@ namespace eosio {
-       * 2. we only have 1 peer so use that.
-       * 3. we have multiple peers, select the next available from the list
+       * a provider is supplied and able to be used, use it.
+       * otherwise select the next available from the list, round-robin style.
        */
 
       if (conn && conn->current() ) {
          source = conn;
       }
-      else if (my_impl->connections.size() == 1) {
-         if (!source) {
-            source = *my_impl->connections.begin();
-         }
-         if (!source->current()) {
-            source.reset();
-         }
-      }
       else {
-         auto cptr = my_impl->connections.begin();
-         auto cend = my_impl->connections.end();
-         if (source) {
-            cptr = my_impl->connections.find(source);
-            cend = cptr;
-            if (cptr == my_impl->connections.end()) {
-               elog ("unable to find previous source connection in connections list");
-               source.reset();
-               cptr = my_impl->connections.begin();
-            } else {
-               if (++cptr == my_impl->connections.end())
-                  cptr = my_impl->connections.begin();
+         if (my_impl->connections.size() == 1) {
+            if (!source) {
+               source = *my_impl->connections.begin();
             }
          }
-         while (cptr != cend) {
-            if ((*cptr)->current()) {
-               source = *cptr;
-               break;
-            }
-            else {
-               if (++cptr == my_impl->connections.end()) {
+         else {
+            // init to a linear array search
+            auto cptr = my_impl->connections.begin();
+            auto cend = my_impl->connections.end();
+            // do we remember the previous source?
+            if (source) {
+               //try to find it in the list
+               cptr = my_impl->connections.find(source);
+               cend = cptr;
+               if (cptr == my_impl->connections.end()) {
+                  //not there - must have been closed! cend is now connections.end, so just flatten the ring.
+                  source.reset();
                   cptr = my_impl->connections.begin();
+               } else {
+                  //was found - advance the start to the next. cend is the old source.
+                  if (++cptr == my_impl->connections.end())
+                     cptr = my_impl->connections.begin();
                }
             }
+
+            //scan the list of peers looking for another able to provide sync blocks.
+            while (cptr != cend) {
+               //select the first one which is current and break out.
+               if ((*cptr)->current()) {
+                  source = *cptr;
+                  break;
+               }
+               else {
+                  // advance the iterator in a round robin fashion.
+                  if (++cptr == my_impl->connections.end()) {
+                     cptr = my_impl->connections.begin();
```

Comment:
What guarantees do we have that this loop will terminate? It seems if the source is not found (and cend is equal to the collection end()) and no connections are "current", then the loop will never end.